### PR TITLE
Nft

### PR DESCRIPTION
**Pull request by alpe**
NFT feature branch.

Replaces : https://github.com/iov-one/weave/pull/120

### CODE REVIEW

File: x/nft/codec.proto [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    string action = 1;
+    repeated Approval approvals = 2;
+}
+
+message Approval {
+    bytes address = 1;
+    ApprovalOptions options = 2;
+}
+
+message ApprovalOptions {
```
Comment (ethanfrey) [inline x/nft/codec.proto]: :+1:

default values `(0, 0, false)` means never expire, but erased on transfer of ownership? Seems like a good default

File: x/nft/model.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+)
+
+func (m *NonFungibleToken) Validate() error {
+	if len(m.Id) < minIDLength || len(m.Id) > maxIDLength {
+		return ErrInvalidID()
+	}
+	if err := weave.Address(m.Owner).Validate(); err != nil {
+		return err
+	}
+	// TODO: impl proper validation
```
Comment (ethanfrey) [inline x/nft/model.go]: This is what #126 is about?

File: x/nft/username/handler.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+)
+
+const (
+	createUsernameCost = 0
+)
+
+// RegisterRoutes will instantiate and register all handlers in this package
+func RegisterRoutes(r weave.Registry, auth x.Authenticator, issuer weave.Address) {
+	bucket := NewBucket()
+	r.Handle(pathIssueTokenMsg, NewIssueHandler(auth, issuer, bucket))
```
Comment (ethanfrey) [inline x/nft/username/handler.go]: no update token?

File: x/nft/model.go (hunk 2 above)
Comment (ruseinov) [reply x/nft/model.go]: Yes, so basically what I wanted to do there is to instantiate ops, pass the token "type",  which we don't have yet and be able to validate approvals fully in the base nft (here).

Alex's suggestion was to split approval validation in several parts and validate it part here and part in impl token.

Alex's approach seems good if we'd need some different logic there, which we don't atm as well as not knowing about type-specific action constants in Approval ops. My argument for my approach was that if we have this centralised - we can validate everything in base nft + make sure we throw an error if we don't know a type.

**Comment by ethanfrey** [comment]
Looks like this is getting ready to merge.... Time for a big rebase!

Let's clarify the merge status and feature status at Tuesday meeting.

**Comment by ruseinov** [comment]
killed that branch with a force-push, reopened in a new PR

**Pull request closed without merging**